idseq_dag/steps/run_star.py

```diff
@@ -1,12 +1,218 @@
+        # Setup
+        input_files = self.input_files_local[0][0:2]
+        star_genome = self.additional_files["star_genome"]
+        ref_dir = self.ref_dir_local
+        scratch_dir = os.path.join(self.output_dir_local, "scratch")
+        total_counts_from_star = {}
+        output_files_local = self.output_files_local()
+        num_inputs = len(self.input_files[0])
+
+        # Check genome file. Ex: s3://...STAR_genome.tar -> /ref/STAR_genome
+        basename = os.path.basename(star_genome).rstrip(".tar")
+        genome_dir = os.path.join(ref_dir, basename)
+        assert os.path.exists(genome_dir)
+        assert genome_dir is not None
+
+        # Check parts file for the number of partitioned indexes
+        parts_file = os.path.join(genome_dir, "parts.txt")
+        assert os.path.isfile(parts_file)
+        with open(parts_file, 'rb') as parts_f:
+            num_parts = int(parts_f.read())
+
+        # Run STAR on each partition and save the unmapped read info
+        unmapped = input_files
+        for part_idx in range(num_parts):
+            tmp = f"{scratch_dir}/star-part-{part_idx}"
+            genome_part = f"{genome_dir}/part-{part_idx}"
+            count_genes = part_idx == 0
+            self.run_star_part(tmp, genome_part, unmapped, count_genes)
+
+            unmapped = self.sync_pairs(self.unmapped_files_in(tmp, num_inputs))
+
+            # Run part 0 in gene-counting mode:
+            # (a) ERCCs are doped into part 0 and we want their counts.
+            # (b) If there is only 1 part (e.g. human), the host gene counts also
+            # make sense.
+            # (c) At part 0, we can also extract out total input reads and if the
+            # total_counts is exactly the same as max_input_reads then we know the
+            # input file is truncated.
+            if part_idx == 0:
+                gene_count_file = os.path.join(tmp, "ReadsPerGene.out.tab")
+                self.extract_total_counts(tmp, num_inputs,
+                                          total_counts_from_star)
+                if os.path.isfile(gene_count_file):
+                    basename = os.path.basename(gene_count_file)
+                    moved = os.path.join(self.output_dir_local, basename)
+                    command.execute(f"mv {gene_count_file} {moved}")
+                    self.additional_files_to_upload.append(moved)
+
+        # Cleanup
+        for src, dst in zip(unmapped, output_files_local):
+            command.execute(f"mv {src} {dst}")  # Move out of scratch dir
+        command.execute("cd %s; rm -rf *" % scratch_dir)
+
+    def run_star_part(self,
+                      output_dir,
+                      genome_dir,
+                      input_files,
+                      count_genes=False):
+        command.execute("mkdir -p %s" % output_dir)
+        cpus = str(multiprocessing.cpu_count())
+        params = [
+            'cd', output_dir, ';', 'STAR', '--outFilterMultimapNmax', '99999',
```

Comment:
Cant this folder be sent on the `cwd` argument (which should in term be exposed by `command.execute`?

Reply:
it seems to lose the cd change in the subprocesses for some reason